// make-plural/pluralCategories.mjs
const z = "zero", o = "one", t = "two", f = "few", m = "many", x = "other";
const a = {cardinal:[o,x],ordinal:[x]};
const b = {cardinal:[x],ordinal:[x]};
const c = {cardinal:[o,f,m,x],ordinal:[x]};
const d = {cardinal:[o,x],ordinal:[o,x]};
const e = {cardinal:[o,t,x],ordinal:[x]};

export const _in = b;
export const af = a;
export const ak = a;
export const am = a;
export const an = a;
export const ar = {cardinal:[z,o,t,f,m,x],ordinal:[x]};
export const ars = {cardinal:[z,o,t,f,m,x],ordinal:[x]};
export const as = {cardinal:[o,x],ordinal:[o,t,f,m,x]};
export const asa = a;
export const ast = a;
export const az = {cardinal:[o,x],ordinal:[o,f,m,x]};
export const be = {cardinal:[o,f,m,x],ordinal:[f,x]};
export const bem = a;
export const bez = a;
export const bg = a;
export const bho = a;
export const bm = b;
export const bn = {cardinal:[o,x],ordinal:[o,t,f,m,x]};
export const bo = b;
export const br = {cardinal:[o,t,f,m,x],ordinal:[x]};
export const brx = a;
export const bs = {cardinal:[o,f,x],ordinal:[x]};
export const ca = {cardinal:[o,x],ordinal:[o,t,f,x]};
export const ce = a;
export const ceb = a;
export const cgg = a;
export const chr = a;
export const ckb = a;
export const cs = c;
export const cy = {cardinal:[z,o,t,f,m,x],ordinal:[z,o,t,f,m,x]};
export const da = a;
export const de = a;
export const dsb = {cardinal:[o,t,f,x],ordinal:[x]};
export const dv = a;
export const dz = b;
export const ee = a;
export const el = a;
export const en = {cardinal:[o,x],ordinal:[o,t,f,x]};
export const eo = a;
export const es = a;
export const et = a;
export const eu = a;
export const fa = a;
export const ff = a;
export const fi = a;
export const fil = d;
export const fo = a;
export const fr = d;
export const fur = a;
export const fy = a;
export const ga = {cardinal:[o,t,f,m,x],ordinal:[o,x]};
export const gd = {cardinal:[o,t,f,x],ordinal:[o,t,f,x]};
export const gl = a;
export const gsw = a;
export const gu = {cardinal:[o,x],ordinal:[o,t,f,m,x]};
export const guw = a;
export const gv = {cardinal:[o,t,f,m,x],ordinal:[x]};
export const ha = a;
export const haw = a;
export const he = {cardinal:[o,t,m,x],ordinal:[x]};
export const hi = {cardinal:[o,x],ordinal:[o,t,f,m,x]};
export const hr = {cardinal:[o,f,x],ordinal:[x]};
export const hsb = {cardinal:[o,t,f,x],ordinal:[x]};
export const hu = d;
export const hy = d;
export const ia = a;
export const id = b;
export const ig = b;
export const ii = b;
export const io = a;
export const is = a;
export const it = {cardinal:[o,x],ordinal:[m,x]};
export const iu = e;
export const iw = {cardinal:[o,t,m,x],ordinal:[x]};
export const ja = b;
export const jbo = b;
export const jgo = a;
export const ji = a;
export const jmc = a;
export const jv = b;
export const jw = b;
export const ka = {cardinal:[o,x],ordinal:[o,m,x]};
export const kab = a;
export const kaj = a;
export const kcg = a;
export const kde = b;
export const kea = b;
export const kk = {cardinal:[o,x],ordinal:[m,x]};
export const kkj = a;
export const kl = a;
export const km = b;
export const kn = a;
export const ko = b;
export const ks = a;
export const ksb = a;
export const ksh = {cardinal:[z,o,x],ordinal:[x]};
export const ku = a;
export const kw = {cardinal:[z,o,t,f,m,x],ordinal:[o,m,x]};
export const ky = a;
export const lag = {cardinal:[z,o,x],ordinal:[x]};
export const lb = a;
export const lg = a;
export const lkt = b;
export const ln = a;
export const lo = {cardinal:[x],ordinal:[o,x]};
export const lt = c;
export const lv = {cardinal:[z,o,x],ordinal:[x]};
export const mas = a;
export const mg = a;
export const mgo = a;
export const mk = {cardinal:[o,x],ordinal:[o,t,m,x]};
export const ml = a;
export const mn = a;
export const mo = {cardinal:[o,f,x],ordinal:[o,x]};
export const mr = {cardinal:[o,x],ordinal:[o,t,f,x]};
export const ms = {cardinal:[x],ordinal:[o,x]};
export const mt = c;
export const my = b;
export const nah = a;
export const naq = e;
export const nb = a;
export const nd = a;
export const ne = d;
export const nl = a;
export const nn = a;
export const nnh = a;
export const no = a;
export const nqo = b;
export const nr = a;
export const nso = a;
export const ny = a;
export const nyn = a;
export const om = a;
export const or = {cardinal:[o,x],ordinal:[o,t,f,m,x]};
export const os = a;
export const osa = b;
export const pa = a;
export const pap = a;
export const pl = c;
export const prg = {cardinal:[z,o,x],ordinal:[x]};
export const ps = a;
export const pt = a;
export const pt_PT = a;
export const rm = a;
export const ro = {cardinal:[o,f,x],ordinal:[o,x]};
export const rof = a;
export const root = b;
export const ru = c;
export const rwk = a;
export const sah = b;
export const saq = a;
export const sc = {cardinal:[o,x],ordinal:[m,x]};
export const scn = {cardinal:[o,x],ordinal:[m,x]};
export const sd = a;
export const sdh = a;
export const se = e;
export const seh = a;
export const ses = b;
export const sg = b;
export const sh = {cardinal:[o,f,x],ordinal:[x]};
export const shi = {cardinal:[o,f,x],ordinal:[x]};
export const si = a;
export const sk = c;
export const sl = {cardinal:[o,t,f,x],ordinal:[x]};
export const sma = e;
export const smi = e;
export const smj = e;
export const smn = e;
export const sms = e;
export const sn = a;
export const so = a;
export const sq = {cardinal:[o,x],ordinal:[o,m,x]};
export const sr = {cardinal:[o,f,x],ordinal:[x]};
export const ss = a;
export const ssy = a;
export const st = a;
export const su = b;
export const sv = d;
export const sw = a;
export const syr = a;
export const ta = a;
export const te = a;
export const teo = a;
export const th = b;
export const ti = a;
export const tig = a;
export const tk = {cardinal:[o,x],ordinal:[f,x]};
export const tl = d;
export const tn = a;
export const to = b;
export const tr = a;
export const ts = a;
export const tzm = a;
export const ug = a;
export const uk = {cardinal:[o,f,m,x],ordinal:[f,x]};
export const ur = a;
export const uz = a;
export const ve = a;
export const vi = {cardinal:[x],ordinal:[o,x]};
export const vo = a;
export const vun = a;
export const wa = a;
export const wae = a;
export const wo = b;
export const xh = a;
export const xog = a;
export const yi = a;
export const yo = b;
export const yue = b;
export const zh = b;
export const zu = a;
