// @foo-software/lighthouse-check/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TRIGGER_TYPE = exports.SUCCESS_CODE_GENERIC = exports.SOURCE_CIRCLECI_ORB = exports.SOURCE_GITHUB_ACTION = exports.DEFAULT_TAG = exports.NAME_STATUS = exports.NAME_RESULTS_JSON_FILE = exports.NAME = exports.API_URL = exports.LIGHTHOUSE_API_URL = exports.FETCH_POLL_INTERVAL = exports.FETCH_POLL_INTERVAL_SECONDS = exports.DEFAULT_FETCH_AND_WAIT_TIMEOUT_MINUTES = exports.API_QUEUE_ITEMS_PATH = exports.API_PAGES_PATH = exports.API_LIGHTHOUSE_AUDIT_PATH = void 0;
// non-alphabetical because some constants depend on others :(
const API_LIGHTHOUSE_AUDIT_PATH = '/lighthouseAudits/queueIds';
exports.API_LIGHTHOUSE_AUDIT_PATH = API_LIGHTHOUSE_AUDIT_PATH;
const API_PAGES_PATH = '/pages';
exports.API_PAGES_PATH = API_PAGES_PATH;
const API_QUEUE_ITEMS_PATH = '/queue/items'; // 10 minutes

exports.API_QUEUE_ITEMS_PATH = API_QUEUE_ITEMS_PATH;
const DEFAULT_FETCH_AND_WAIT_TIMEOUT_MINUTES = 10; // 10 seconds

exports.DEFAULT_FETCH_AND_WAIT_TIMEOUT_MINUTES = DEFAULT_FETCH_AND_WAIT_TIMEOUT_MINUTES;
const FETCH_POLL_INTERVAL_SECONDS = 10;
exports.FETCH_POLL_INTERVAL_SECONDS = FETCH_POLL_INTERVAL_SECONDS;
const FETCH_POLL_INTERVAL = FETCH_POLL_INTERVAL_SECONDS * 1000;
exports.FETCH_POLL_INTERVAL = FETCH_POLL_INTERVAL;
const LIGHTHOUSE_API_URL = 'https://www.foo.software/api/v1';
exports.LIGHTHOUSE_API_URL = LIGHTHOUSE_API_URL;
const API_URL = process.env.API_URL || LIGHTHOUSE_API_URL;
exports.API_URL = API_URL;
const NAME = 'lighthouse-check';
exports.NAME = NAME;
const NAME_RESULTS_JSON_FILE = 'results.json';
exports.NAME_RESULTS_JSON_FILE = NAME_RESULTS_JSON_FILE;
const NAME_STATUS = 'lighthouse-check-status';
exports.NAME_STATUS = NAME_STATUS;
const DEFAULT_TAG = NAME;
exports.DEFAULT_TAG = DEFAULT_TAG;
const SOURCE_GITHUB_ACTION = 'GitHub Action';
exports.SOURCE_GITHUB_ACTION = SOURCE_GITHUB_ACTION;
const SOURCE_CIRCLECI_ORB = 'CircleCI Orb';
exports.SOURCE_CIRCLECI_ORB = SOURCE_CIRCLECI_ORB;
const SUCCESS_CODE_GENERIC = 'SUCCESS';
exports.SUCCESS_CODE_GENERIC = SUCCESS_CODE_GENERIC;
const TRIGGER_TYPE = 'lighthouseAudit';
exports.TRIGGER_TYPE = TRIGGER_TYPE;
