// @foo-software/lighthouse-check/dist/errorCodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ERROR_QUEUE_MAX_USED_DAY = exports.ERROR_UNAUTHORIZED = exports.ERROR_UNEXPECTED_RESPONSE = exports.ERROR_TIMEOUT = exports.ERROR_NO_URLS = exports.ERROR_NO_RESULTS = exports.ERROR_INVALID = exports.ERROR_GENERIC = void 0;
const ERROR_GENERIC = 'ERROR_GENERIC';
exports.ERROR_GENERIC = ERROR_GENERIC;
const ERROR_INVALID = 'ERROR_INVALID';
exports.ERROR_INVALID = ERROR_INVALID;
const ERROR_NO_RESULTS = 'ERROR_NO_RESULTS';
exports.ERROR_NO_RESULTS = ERROR_NO_RESULTS;
const ERROR_NO_URLS = 'ERROR_NO_URLS';
exports.ERROR_NO_URLS = ERROR_NO_URLS;
const ERROR_TIMEOUT = 'ERROR_TIMEOUT';
exports.ERROR_TIMEOUT = ERROR_TIMEOUT;
const ERROR_UNEXPECTED_RESPONSE = 'ERROR_UNEXPECTED_RESPONSE';
exports.ERROR_UNEXPECTED_RESPONSE = ERROR_UNEXPECTED_RESPONSE;
const ERROR_UNAUTHORIZED = 'ERROR_UNAUTHORIZED'; // from API

exports.ERROR_UNAUTHORIZED = ERROR_UNAUTHORIZED;
const ERROR_QUEUE_MAX_USED_DAY = 'ERROR_QUEUE_MAX_USED_DAY';
exports.ERROR_QUEUE_MAX_USED_DAY = ERROR_QUEUE_MAX_USED_DAY;
