// @foo-software/lighthouse-check/dist/fetchLighthouseAudits.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash.get"));

var _fetch = _interopRequireDefault(require("./fetch"));

var _LighthouseCheckError = _interopRequireDefault(require("./LighthouseCheckError"));

var _constants = require("./constants");

var _errorCodes = require("./errorCodes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = async ({
  apiToken,
  queueIds
}) => {
  try {
    const lighthouseAuditsResponse = await (0, _fetch.default)(`${_constants.API_URL}${_constants.API_LIGHTHOUSE_AUDIT_PATH}/${queueIds.join()}`, {
      method: 'get',
      headers: {
        Authorization: apiToken,
        'Content-Type': 'application/json'
      }
    });
    const lighthouseAuditsJson = await lighthouseAuditsResponse.json();

    if (lighthouseAuditsJson.status >= 400) {
      throw new _LighthouseCheckError.default('No results found.', {
        code: _errorCodes.ERROR_NO_RESULTS
      });
    }

    const lighthouseResults = (0, _lodash.default)(lighthouseAuditsJson, 'data.lighthouseaudit', []); // success

    return {
      code: _constants.SUCCESS_CODE_GENERIC,
      data: lighthouseResults,
      message: 'Lighthouse results successfully fetched.'
    };
  } catch (error) {
    return {
      code: error.code || _errorCodes.ERROR_GENERIC,
      error
    };
  }
};

exports.default = _default;
