// @foo-software/lighthouse-check/dist/postPrComment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fetch = _interopRequireDefault(require("./fetch"));

var _getLighthouseScoreColor = _interopRequireDefault(require("./helpers/getLighthouseScoreColor"));

var _lighthouseAuditTitles = _interopRequireDefault(require("./lighthouseAuditTitles"));

var _LighthouseCheckError = _interopRequireDefault(require("./LighthouseCheckError"));

var _errorCodes = require("./errorCodes");

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getBadge = ({
  title,
  score
}) => `![](https://img.shields.io/badge/${title}-${score}-${(0, _getLighthouseScoreColor.default)({
  isHex: false,
  score
})}?style=flat-square) `;

var _default = async ({
  prCommentAccessToken,
  prCommentSaveOld,
  prCommentUrl,
  results,
  verbose
}) => {
  try {
    let markdown = '';
    results.forEach((result, index) => {
      // badges
      Object.keys(result.scores).forEach(current => {
        markdown += getBadge({
          title: _lighthouseAuditTitles.default[current].replace(/ /g, '%20'),
          score: result.scores[current]
        });
      }); // the url

      markdown += `\n\n${result.url}`; // if we have a URL for the full report

      if (result.report) {
        markdown += `\n\n[Full report](${result.report})`;
      } // add a horizontal line


      if (index + 1 < results.length) {
        markdown += `\n\n<hr />\n\n`;
      }
    }); // create an identifier within the comment when searching comments
    // in the future

    const commentIdentifier = '<!-- generated by lighthouse-check -->';
    markdown += commentIdentifier; // establish existing comment

    let existingComment; // if we aren't saving old comments

    if (!prCommentSaveOld) {
      // get existing comments to see if we've already posted scores
      const existingCommentsResult = await (0, _fetch.default)(prCommentUrl, {
        method: 'get',
        headers: {
          'content-type': 'application/json',
          authorization: `token ${prCommentAccessToken}`
        }
      });
      const existingCommentsJsonResult = await existingCommentsResult.json();

      if (Array.isArray(existingCommentsJsonResult) && existingCommentsJsonResult.length) {
        existingComment = existingCommentsJsonResult.find(current => current.body.includes(commentIdentifier));
      }
    } // create or update the comment with scores


    const shouldUpdate = existingComment && existingComment.id;
    const url = !shouldUpdate ? prCommentUrl : `${prCommentUrl}/${existingComment.id}`;
    const result = await (0, _fetch.default)(url, {
      method: !shouldUpdate ? 'post' : 'put',
      body: JSON.stringify({ ...(shouldUpdate ? {} : {
          event: 'COMMENT'
        }),
        body: markdown
      }),
      headers: {
        'content-type': 'application/json',
        authorization: `token ${prCommentAccessToken}`
      }
    });
    const jsonResult = await result.json();

    if (!jsonResult.id) {
      throw new _LighthouseCheckError.default(jsonResult.message || 'something went wrong', {
        code: _errorCodes.ERROR_UNEXPECTED_RESPONSE,
        data: jsonResult
      });
    }
  } catch (error) {
    if (verbose) {
      console.log(`${_constants.NAME}:`, error);
    } // we still need to kill the process


    throw error;
  }
};

exports.default = _default;
