// lighthouse/report/renderer/components.js
'use strict';

// auto-generated by build/build-report-components.js

/** @typedef {import('./dom.js').DOM} DOM */

/* eslint-disable max-len */


/**
 * @param {DOM} dom
 */
function create3pFilterComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    .lh-3p-filter {\n      background-color: var(--table-higlight-background-color);\n      color: var(--color-gray-600);\n      float: right;\n      padding: 6px;\n    }\n    .lh-3p-filter-label, .lh-3p-filter-input {\n      vertical-align: middle;\n      user-select: none;\n    }\n    .lh-3p-filter-input:disabled + .lh-3p-ui-string {\n      text-decoration: line-through;\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('div', 'lh-3p-filter');
  const el3 = dom.createElement('label', 'lh-3p-filter-label');
  const el4 = dom.createElement('input', 'lh-3p-filter-input');
  el4.setAttribute('type', 'checkbox');
  el4.setAttribute('checked', '');
  const el5 = dom.createElement('span', 'lh-3p-ui-string');
  el5.append('Show 3rd party resources');
  const el6 = dom.createElement('span', 'lh-3p-filter-count');
  el3.append(' ', el4, ' ', el5, ' (', el6, ') ');
  el2.append(' ', el3, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createAuditComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-audit');
  const el2 = dom.createElement('details', 'lh-expandable-details');
  const el3 = dom.createElement('summary');
  const el4 = dom.createElement('div', 'lh-audit__header lh-expandable-details__summary');
  const el5 = dom.createElement('span', 'lh-audit__score-icon');
  const el6 = dom.createElement('span', 'lh-audit__title-and-text');
  const el7 = dom.createElement('span', 'lh-audit__title');
  const el8 = dom.createElement('span', 'lh-audit__display-text');
  el6.append(' ', el7, ' ', el8, ' ');
  const el9 = dom.createElement('div', 'lh-chevron-container');
  el4.append(' ', el5, ' ', el6, ' ', el9, ' ');
  el3.append(' ', el4, ' ');
  const el10 = dom.createElement('div', 'lh-audit__description');
  const el11 = dom.createElement('div', 'lh-audit__stackpacks');
  el2.append(' ', el3, ' ', el10, ' ', el11, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createCategoryHeaderComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-category-header');
  const el2 = dom.createElement('div', 'lh-score__gauge');
  el2.setAttribute('role', 'heading');
  el2.setAttribute('aria-level', '2');
  const el3 = dom.createElement('div', 'lh-category-header__description');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createChevronComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg', 'lh-chevron');
  el1.setAttribute('viewBox', '0 0 100 100');
  const el2 = dom.createElementNS('http://www.w3.org/2000/svg', 'g', 'lh-chevron__lines');
  const el3 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-chevron__line lh-chevron__line-left');
  el3.setAttribute('d', 'M10 50h40');
  const el4 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-chevron__line lh-chevron__line-right');
  el4.setAttribute('d', 'M90 50H50');
  el2.append(' ', el3, ' ', el4, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createClumpComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('details', 'lh-clump lh-audit-group');
  const el2 = dom.createElement('summary');
  const el3 = dom.createElement('div', 'lh-audit-group__summary');
  const el4 = dom.createElement('div', 'lh-audit-group__header');
  const el5 = dom.createElement('span', 'lh-audit-group__title');
  const el6 = dom.createElement('span', 'lh-audit-group__itemcount');
  el4.append(' ', el5, ' ', el6, ' ', ' ', ' ');
  el3.append(' ', el4, ' ');
  el2.append(' ', el3, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createCrcComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-crc-container');
  const el2 = dom.createElement('style');
  el2.append('\n      .lh-crc .lh-tree-marker {\n        width: 12px;\n        height: 26px;\n        display: block;\n        float: left;\n        background-position: top left;\n      }\n      .lh-crc .lh-horiz-down {\n        background: url(\'data:image/svg+xml;utf8,<svg width="16" height="26" viewBox="0 0 16 26" xmlns="http://www.w3.org/2000/svg"><g fill="%23D8D8D8" fill-rule="evenodd"><path d="M16 12v2H-2v-2z"/><path d="M9 12v14H7V12z"/></g></svg>\');\n      }\n      .lh-crc .lh-right {\n        background: url(\'data:image/svg+xml;utf8,<svg width="16" height="26" viewBox="0 0 16 26" xmlns="http://www.w3.org/2000/svg"><path d="M16 12v2H0v-2z" fill="%23D8D8D8" fill-rule="evenodd"/></svg>\');\n      }\n      .lh-crc .lh-up-right {\n        background: url(\'data:image/svg+xml;utf8,<svg width="16" height="26" viewBox="0 0 16 26" xmlns="http://www.w3.org/2000/svg"><path d="M7 0h2v14H7zm2 12h7v2H9z" fill="%23D8D8D8" fill-rule="evenodd"/></svg>\');\n      }\n      .lh-crc .lh-vert-right {\n        background: url(\'data:image/svg+xml;utf8,<svg width="16" height="26" viewBox="0 0 16 26" xmlns="http://www.w3.org/2000/svg"><path d="M7 0h2v27H7zm2 12h7v2H9z" fill="%23D8D8D8" fill-rule="evenodd"/></svg>\');\n      }\n      .lh-crc .lh-vert {\n        background: url(\'data:image/svg+xml;utf8,<svg width="16" height="26" viewBox="0 0 16 26" xmlns="http://www.w3.org/2000/svg"><path d="M7 0h2v26H7z" fill="%23D8D8D8" fill-rule="evenodd"/></svg>\');\n      }\n      .lh-crc .lh-crc-tree {\n        font-size: 14px;\n        width: 100%;\n        overflow-x: auto;\n      }\n      .lh-crc .lh-crc-node {\n        height: 26px;\n        line-height: 26px;\n        white-space: nowrap;\n      }\n      .lh-crc .lh-crc-node__tree-value {\n        margin-left: 10px;\n      }\n      .lh-crc .lh-crc-node__tree-value div {\n        display: inline;\n      }\n      .lh-crc .lh-crc-node__chain-duration {\n        font-weight: 700;\n      }\n      .lh-crc .lh-crc-initial-nav {\n        color: #595959;\n        font-style: italic;\n      }\n      .lh-crc__summary-value {\n        margin-bottom: 10px;\n      }\n    ');
  const el3 = dom.createElement('div');
  const el4 = dom.createElement('div', 'lh-crc__summary-value');
  const el5 = dom.createElement('span', 'lh-crc__longest_duration_label');
  const el6 = dom.createElement('b', 'lh-crc__longest_duration');
  el4.append(' ', el5, ' ', el6, ' ');
  el3.append(' ', el4, ' ');
  const el7 = dom.createElement('div', 'lh-crc');
  const el8 = dom.createElement('div', 'lh-crc-initial-nav');
  el7.append(' ', el8, ' ', ' ');
  el1.append(' ', el2, ' ', el3, ' ', el7, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createCrcChainComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-crc-node');
  const el2 = dom.createElement('span', 'lh-crc-node__tree-marker');
  const el3 = dom.createElement('span', 'lh-crc-node__tree-value');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createElementScreenshotComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-element-screenshot');
  const el2 = dom.createElement('div', 'lh-element-screenshot__content');
  const el3 = dom.createElement('div', 'lh-element-screenshot__mask');
  const el4 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg');
  el4.setAttribute('height', '0');
  el4.setAttribute('width', '0');
  const el5 = dom.createElementNS('http://www.w3.org/2000/svg', 'defs');
  const el6 = dom.createElementNS('http://www.w3.org/2000/svg', 'clipPath');
  el6.setAttribute('clipPathUnits', 'objectBoundingBox');
  el5.append(' ', el6, ' ', ' ');
  el4.append(' ', el5, ' ');
  el3.append(' ', el4, ' ');
  const el7 = dom.createElement('div', 'lh-element-screenshot__image');
  const el8 = dom.createElement('div', 'lh-element-screenshot__element-marker');
  el2.append(' ', el3, ' ', el7, ' ', el8, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createEnvItemComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('li', 'lh-env__item');
  const el2 = dom.createElement('span', 'lh-env__name');
  const el3 = dom.createElement('span', 'lh-env__description');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createFooterComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    .lh-footer {\n      padding: var(--footer-padding-vertical) calc(var(--default-padding) * 2);\n      max-width: var(--report-width);\n      margin: 0 auto;\n    }\n    .lh-footer .lh-generated {\n      text-align: center;\n    }\n    .lh-env__title {\n      font-size: var(--env-item-font-size-big);\n      line-height: var(--env-item-line-height-big);\n      text-align: center;\n      padding: var(--score-container-padding);\n    }\n    .lh-env {\n      padding: var(--default-padding) 0;\n    }\n    .lh-env__items {\n      padding-left: 16px;\n      margin: 0 0 var(--audits-margin-bottom);\n      padding: 0;\n    }\n    .lh-env__items .lh-env__item:nth-child(2n) {\n      background-color: var(--env-item-background-color);\n    }\n    .lh-env__item {\n      display: flex;\n      padding: var(--env-item-padding);\n      position: relative;\n    }\n    span.lh-env__name {\n      font-weight: bold;\n      min-width: var(--env-name-min-width);\n      flex: 0.5;\n      padding: 0 8px;\n    }\n    span.lh-env__description {\n      text-align: left;\n      flex: 1;\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('footer', 'lh-footer');
  const el3 = dom.createElement('div', 'lh-env');
  const el4 = dom.createElement('div', 'lh-env__title');
  el4.append('Runtime Settings');
  const el5 = dom.createElement('ul', 'lh-env__items');
  el5.append(' ', ' ');
  el3.append(' ', el4, ' ', el5, ' ');
  const el6 = dom.createElement('div', 'lh-generated');
  const el7 = dom.createElement('b');
  el7.append('Lighthouse');
  const el8 = dom.createElement('span', 'lh-footer__version');
  const el9 = dom.createElement('a', 'lh-footer__version_issue');
  el9.setAttribute('href', 'https://github.com/GoogleChrome/Lighthouse/issues');
  el9.setAttribute('target', '_blank');
  el9.setAttribute('rel', 'noopener');
  el9.append('File an issue');
  el6.append(' ', ' Generated by ', el7, ' ', el8, ' | ', el9, ' ');
  el2.append(' ', ' ', el3, ' ', el6, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createFractionComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('a', 'lh-fraction__wrapper');
  el1.setAttribute('href', '#');
  const el2 = dom.createElement('div', 'lh-fraction__content-wrapper');
  const el3 = dom.createElement('div', 'lh-fraction__content');
  const el4 = dom.createElement('div', 'lh-fraction__background');
  el3.append(' ', el4, ' ');
  el2.append(' ', el3, ' ');
  const el5 = dom.createElement('div', 'lh-fraction__label');
  el1.append(' ', el2, ' ', el5, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createGaugeComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('a', 'lh-gauge__wrapper');
  el1.setAttribute('href', '#');
  const el2 = dom.createElement('div', 'lh-gauge__svg-wrapper');
  const el3 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg', 'lh-gauge');
  el3.setAttribute('viewBox', '0 0 120 120');
  const el4 = dom.createElementNS('http://www.w3.org/2000/svg', 'circle', 'lh-gauge-base');
  el4.setAttribute('r', '56');
  el4.setAttribute('cx', '60');
  el4.setAttribute('cy', '60');
  el4.setAttribute('stroke-width', '8');
  const el5 = dom.createElementNS('http://www.w3.org/2000/svg', 'circle', 'lh-gauge-arc');
  el5.setAttribute('r', '56');
  el5.setAttribute('cx', '60');
  el5.setAttribute('cy', '60');
  el5.setAttribute('stroke-width', '8');
  el3.append(' ', el4, ' ', el5, ' ');
  el2.append(' ', el3, ' ');
  const el6 = dom.createElement('div', 'lh-gauge__percentage');
  const el7 = dom.createElement('div', 'lh-gauge__label');
  el1.append(' ', ' ', el2, ' ', el6, ' ', ' ', el7, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createGaugePwaComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    .lh-gauge--pwa .lh-gauge--pwa__component {\n      display: none;\n    }\n    .lh-gauge--pwa__wrapper:not(.lh-badged--all) .lh-gauge--pwa__logo > path {\n      /* Gray logo unless everything is passing. */\n      fill: #B0B0B0;\n    }\n\n    .lh-gauge--pwa__disc {\n      fill: var(--color-gray-200);\n    }\n\n    .lh-gauge--pwa__logo--primary-color {\n      fill: #304FFE;\n    }\n\n    .lh-gauge--pwa__logo--secondary-color {\n      fill: #3D3D3D;\n    }\n    .lh-dark .lh-gauge--pwa__logo--secondary-color {\n      fill: #D8B6B6;\n    }\n\n    /* No passing groups. */\n    .lh-gauge--pwa__wrapper:not([class*=\'lh-badged--\']) .lh-gauge--pwa__na-line {\n      display: inline;\n    }\n    /* Just optimized. Same n/a line as no passing groups. */\n    .lh-gauge--pwa__wrapper.lh-badged--pwa-optimized:not(.lh-badged--pwa-installable) .lh-gauge--pwa__na-line {\n      display: inline;\n    }\n\n    /* Just installable. */\n    .lh-gauge--pwa__wrapper.lh-badged--pwa-installable .lh-gauge--pwa__installable-badge {\n      display: inline;\n    }\n\n    /* All passing groups. */\n    .lh-gauge--pwa__wrapper.lh-badged--all .lh-gauge--pwa__check-circle {\n      display: inline;\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('a', 'lh-gauge__wrapper lh-gauge--pwa__wrapper');
  el2.setAttribute('href', '#');
  const el3 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg', 'lh-gauge lh-gauge--pwa');
  el3.setAttribute('viewBox', '0 0 60 60');
  const el4 = dom.createElementNS('http://www.w3.org/2000/svg', 'defs');
  const el5 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el5.setAttribute('id', 'lh-gauge--pwa__check-circle__gradient');
  el5.setAttribute('x1', '50%');
  el5.setAttribute('y1', '0%');
  el5.setAttribute('x2', '50%');
  el5.setAttribute('y2', '100%');
  const el6 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el6.setAttribute('stop-color', '#00C852');
  el6.setAttribute('offset', '0%');
  const el7 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el7.setAttribute('stop-color', '#009688');
  el7.setAttribute('offset', '100%');
  el5.append(' ', el6, ' ', el7, ' ');
  const el8 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el8.setAttribute('id', 'lh-gauge--pwa__installable__shadow-gradient');
  el8.setAttribute('x1', '76.056%');
  el8.setAttribute('x2', '24.111%');
  el8.setAttribute('y1', '82.995%');
  el8.setAttribute('y2', '24.735%');
  const el9 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el9.setAttribute('stop-color', '#A5D6A7');
  el9.setAttribute('offset', '0%');
  const el10 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el10.setAttribute('stop-color', '#80CBC4');
  el10.setAttribute('offset', '100%');
  el8.append(' ', el9, ' ', el10, ' ');
  const el11 = dom.createElementNS('http://www.w3.org/2000/svg', 'g');
  el11.setAttribute('id', 'lh-gauge--pwa__installable-badge');
  const el12 = dom.createElementNS('http://www.w3.org/2000/svg', 'circle');
  el12.setAttribute('fill', '#FFFFFF');
  el12.setAttribute('cx', '10');
  el12.setAttribute('cy', '10');
  el12.setAttribute('r', '10');
  const el13 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el13.setAttribute('fill', '#009688');
  el13.setAttribute('d', 'M10 4.167A5.835 5.835 0 0 0 4.167 10 5.835 5.835 0 0 0 10 15.833 5.835 5.835 0 0 0 15.833 10 5.835 5.835 0 0 0 10 4.167zm2.917 6.416h-2.334v2.334H9.417v-2.334H7.083V9.417h2.334V7.083h1.166v2.334h2.334v1.166z');
  el11.append(' ', el12, ' ', el13, ' ');
  el4.append(' ', el5, ' ', el8, ' ', el11, ' ');
  const el14 = dom.createElementNS('http://www.w3.org/2000/svg', 'g');
  el14.setAttribute('stroke', 'none');
  el14.setAttribute('fill-rule', 'nonzero');
  const el15 = dom.createElementNS('http://www.w3.org/2000/svg', 'circle', 'lh-gauge--pwa__disc');
  el15.setAttribute('cx', '30');
  el15.setAttribute('cy', '30');
  el15.setAttribute('r', '30');
  const el16 = dom.createElementNS('http://www.w3.org/2000/svg', 'g', 'lh-gauge--pwa__logo');
  const el17 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-gauge--pwa__logo--secondary-color');
  el17.setAttribute('d', 'M35.66 19.39l.7-1.75h2L37.4 15 38.6 12l3.4 9h-2.51l-.58-1.61z');
  const el18 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-gauge--pwa__logo--primary-color');
  el18.setAttribute('d', 'M33.52 21l3.65-9h-2.42l-2.5 5.82L30.5 12h-1.86l-1.9 5.82-1.35-2.65-1.21 3.72L25.4 21h2.38l1.72-5.2 1.64 5.2z');
  const el19 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-gauge--pwa__logo--secondary-color');
  el19.setAttribute('fill-rule', 'nonzero');
  el19.setAttribute('d', 'M20.3 17.91h1.48c.45 0 .85-.05 1.2-.15l.39-1.18 1.07-3.3a2.64 2.64 0 0 0-.28-.37c-.55-.6-1.36-.91-2.42-.91H18v9h2.3V17.9zm1.96-3.84c.22.22.33.5.33.87 0 .36-.1.65-.29.87-.2.23-.59.35-1.15.35h-.86v-2.41h.87c.52 0 .89.1 1.1.32z');
  el16.append(' ', el17, ' ', el18, ' ', el19, ' ');
  const el20 = dom.createElementNS('http://www.w3.org/2000/svg', 'rect', 'lh-gauge--pwa__component lh-gauge--pwa__na-line');
  el20.setAttribute('fill', '#FFFFFF');
  el20.setAttribute('x', '20');
  el20.setAttribute('y', '32');
  el20.setAttribute('width', '20');
  el20.setAttribute('height', '4');
  el20.setAttribute('rx', '2');
  const el21 = dom.createElementNS('http://www.w3.org/2000/svg', 'g', 'lh-gauge--pwa__component lh-gauge--pwa__installable-badge');
  el21.setAttribute('transform', 'translate(20, 29)');
  const el22 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el22.setAttribute('fill', 'url(#lh-gauge--pwa__installable__shadow-gradient)');
  el22.setAttribute('d', 'M33.629 19.487c-4.272 5.453-10.391 9.39-17.415 10.869L3 17.142 17.142 3 33.63 19.487z');
  const el23 = dom.createElementNS('http://www.w3.org/2000/svg', 'use');
  el23.setAttribute('href', '#lh-gauge--pwa__installable-badge');
  el21.append(' ', el22, ' ', el23, ' ');
  const el24 = dom.createElementNS('http://www.w3.org/2000/svg', 'g', 'lh-gauge--pwa__component lh-gauge--pwa__check-circle');
  el24.setAttribute('transform', 'translate(18, 28)');
  const el25 = dom.createElementNS('http://www.w3.org/2000/svg', 'circle');
  el25.setAttribute('fill', '#FFFFFF');
  el25.setAttribute('cx', '12');
  el25.setAttribute('cy', '12');
  el25.setAttribute('r', '12');
  const el26 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el26.setAttribute('fill', 'url(#lh-gauge--pwa__check-circle__gradient)');
  el26.setAttribute('d', 'M12 2a10 10 0 1 0 0 20 10 10 0 0 0 0-20zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z');
  el24.append(' ', el25, ' ', el26, ' ');
  el14.append(' ', ' ', el15, ' ', el16, ' ', ' ', el20, ' ', ' ', el21, ' ', ' ', el24, ' ');
  el3.append(' ', el4, ' ', el14, ' ');
  const el27 = dom.createElement('div', 'lh-gauge__label');
  el2.append(' ', el3, ' ', el27, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createHeadingComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    /* CSS Fireworks. Originally by Eddie Lin\n       https://codepen.io/paulirish/pen/yEVMbP\n    */\n    .lh-pyro {\n      display: none;\n      z-index: 1;\n      pointer-events: none;\n    }\n    .lh-score100 .lh-pyro {\n      display: block;\n    }\n    .lh-score100 .lh-lighthouse stop:first-child {\n      stop-color: hsla(200, 12%, 95%, 0);\n    }\n    .lh-score100 .lh-lighthouse stop:last-child {\n      stop-color: hsla(65, 81%, 76%, 1);\n    }\n\n    .lh-pyro > .lh-pyro-before, .lh-pyro > .lh-pyro-after {\n      position: absolute;\n      width: 5px;\n      height: 5px;\n      border-radius: 2.5px;\n      box-shadow: 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff, 0 0 #fff;\n      animation: 1s bang ease-out infinite backwards,  1s gravity ease-in infinite backwards,  5s position linear infinite backwards;\n      animation-delay: 1s, 1s, 1s;\n    }\n\n    .lh-pyro > .lh-pyro-after {\n      animation-delay: 2.25s, 2.25s, 2.25s;\n      animation-duration: 1.25s, 1.25s, 6.25s;\n    }\n    .lh-fireworks-paused .lh-pyro > div {\n      animation-play-state: paused;\n    }\n\n    @keyframes bang {\n      to {\n        box-shadow: -70px -115.67px #47ebbc, -28px -99.67px #eb47a4, 58px -31.67px #7eeb47, 13px -141.67px #eb47c5, -19px 6.33px #7347eb, -2px -74.67px #ebd247, 24px -151.67px #eb47e0, 57px -138.67px #b4eb47, -51px -104.67px #479eeb, 62px 8.33px #ebcf47, -93px 0.33px #d547eb, -16px -118.67px #47bfeb, 53px -84.67px #47eb83, 66px -57.67px #eb47bf, -93px -65.67px #91eb47, 30px -13.67px #86eb47, -2px -59.67px #83eb47, -44px 1.33px #eb47eb, 61px -58.67px #47eb73, 5px -22.67px #47e8eb, -66px -28.67px #ebe247, 42px -123.67px #eb5547, -75px 26.33px #7beb47, 15px -52.67px #a147eb, 36px -51.67px #eb8347, -38px -12.67px #eb5547, -46px -59.67px #47eb81, 78px -114.67px #eb47ba, 15px -156.67px #eb47bf, -36px 1.33px #eb4783, -72px -86.67px #eba147, 31px -46.67px #ebe247, -68px 29.33px #47e2eb, -55px 19.33px #ebe047, -56px 27.33px #4776eb, -13px -91.67px #eb5547, -47px -138.67px #47ebc7, -18px -96.67px #eb47ac, 11px -88.67px #4783eb, -67px -28.67px #47baeb, 53px 10.33px #ba47eb, 11px 19.33px #5247eb, -5px -11.67px #eb4791, -68px -4.67px #47eba7, 95px -37.67px #eb478b, -67px -162.67px #eb5d47, -54px -120.67px #eb6847, 49px -12.67px #ebe047, 88px 8.33px #47ebda, 97px 33.33px #eb8147, 6px -71.67px #ebbc47;\n      }\n    }\n    @keyframes gravity {\n      to {\n        transform: translateY(80px);\n        opacity: 0;\n      }\n    }\n    @keyframes position {\n      0%, 19.9% {\n        margin-top: 4%;\n        margin-left: 47%;\n      }\n      20%, 39.9% {\n        margin-top: 7%;\n        margin-left: 30%;\n      }\n      40%, 59.9% {\n        margin-top: 6%;\n        margin-left: 70%;\n      }\n      60%, 79.9% {\n        margin-top: 3%;\n        margin-left: 20%;\n      }\n      80%, 99.9% {\n        margin-top: 3%;\n        margin-left: 80%;\n      }\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('div', 'lh-header-container');
  const el3 = dom.createElement('div', 'lh-scores-wrapper-placeholder');
  el2.append(' ', el3, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createMetricComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-metric');
  const el2 = dom.createElement('div', 'lh-metric__innerwrap');
  const el3 = dom.createElement('div', 'lh-metric__icon');
  const el4 = dom.createElement('span', 'lh-metric__title');
  const el5 = dom.createElement('div', 'lh-metric__value');
  const el6 = dom.createElement('div', 'lh-metric__description');
  el2.append(' ', el3, ' ', el4, ' ', el5, ' ', el6, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createMetricsToggleComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-metrics-toggle');
  const el2 = dom.createElement('input', 'lh-metrics-toggle__input');
  el2.setAttribute('type', 'checkbox');
  el2.setAttribute('id', 'toggle-metric-descriptions');
  el2.setAttribute('aria-label', 'Toggle the display of metric descriptions');
  const el3 = dom.createElement('label', 'lh-metrics-toggle__label');
  el3.setAttribute('for', 'toggle-metric-descriptions');
  const el4 = dom.createElement('div', 'lh-metrics-toggle__icon lh-metrics-toggle__icon--less');
  el4.setAttribute('aria-hidden', 'true');
  const el5 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg');
  el5.setAttribute('width', '24');
  el5.setAttribute('height', '24');
  el5.setAttribute('viewBox', '0 0 24 24');
  const el6 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-metrics-toggle__lines');
  el6.setAttribute('d', 'M4 9h16v2H4zm0 4h10v2H4z');
  el5.append(' ', el6, ' ');
  el4.append(' ', el5, ' ');
  const el7 = dom.createElement('div', 'lh-metrics-toggle__icon lh-metrics-toggle__icon--more');
  el7.setAttribute('aria-hidden', 'true');
  const el8 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg');
  el8.setAttribute('width', '24');
  el8.setAttribute('height', '24');
  el8.setAttribute('viewBox', '0 0 24 24');
  const el9 = dom.createElementNS('http://www.w3.org/2000/svg', 'path', 'lh-metrics-toggle__lines');
  el9.setAttribute('d', 'M3 18h12v-2H3v2zM3 6v2h18V6H3zm0 7h18v-2H3v2z');
  el8.append(' ', el9, ' ');
  el7.append(' ', el8, ' ');
  el3.append(' ', el4, ' ', el7, ' ');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createOpportunityComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-audit lh-audit--load-opportunity');
  const el2 = dom.createElement('details', 'lh-expandable-details');
  const el3 = dom.createElement('summary');
  const el4 = dom.createElement('div', 'lh-audit__header lh-expandable-details__summary');
  const el5 = dom.createElement('div', 'lh-load-opportunity__cols');
  const el6 = dom.createElement('div', 'lh-load-opportunity__col lh-load-opportunity__col--one');
  const el7 = dom.createElement('span', 'lh-audit__score-icon');
  const el8 = dom.createElement('div', 'lh-audit__title');
  el6.append(' ', el7, ' ', el8, ' ');
  const el9 = dom.createElement('div', 'lh-load-opportunity__col lh-load-opportunity__col--two');
  const el10 = dom.createElement('div', 'lh-load-opportunity__sparkline');
  const el11 = dom.createElement('div', 'lh-sparkline');
  const el12 = dom.createElement('div', 'lh-sparkline__bar');
  el11.append(el12);
  el10.append(' ', el11, ' ');
  const el13 = dom.createElement('div', 'lh-audit__display-text');
  const el14 = dom.createElement('div', 'lh-chevron-container');
  el9.append(' ', el10, ' ', el13, ' ', el14, ' ');
  el5.append(' ', el6, ' ', el9, ' ');
  el4.append(' ', el5, ' ');
  el3.append(' ', el4, ' ');
  const el15 = dom.createElement('div', 'lh-audit__description');
  const el16 = dom.createElement('div', 'lh-audit__stackpacks');
  el2.append(' ', el3, ' ', el15, ' ', el16, ' ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createOpportunityHeaderComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-load-opportunity__header lh-load-opportunity__cols');
  const el2 = dom.createElement('div', 'lh-load-opportunity__col lh-load-opportunity__col--one');
  const el3 = dom.createElement('div', 'lh-load-opportunity__col lh-load-opportunity__col--two');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createScorescaleComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-scorescale');
  const el2 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--fail');
  el2.append('0–49');
  const el3 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--average');
  el3.append('50–89');
  const el4 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--pass');
  el4.append('90–100');
  el1.append(' ', el2, ' ', el3, ' ', el4, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createScoresWrapperComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    .lh-scores-container {\n      display: flex;\n      flex-direction: column;\n      padding: var(--scores-container-padding);\n      position: relative;\n      width: 100%;\n    }\n\n    .lh-sticky-header {\n      --gauge-circle-size: 36px;\n      --plugin-badge-size: 18px;\n      --plugin-icon-size: 75%;\n      --gauge-wrapper-width: 60px;\n      --gauge-percentage-font-size: 13px;\n      position: fixed;\n      left: 0;\n      right: 0;\n      top: var(--topbar-height);\n      font-weight: 700;\n      display: none;\n      justify-content: center;\n      background-color: var(--sticky-header-background-color);\n      border-bottom: 1px solid var(--color-gray-200);\n      padding-top: var(--score-container-padding);\n      padding-bottom: 4px;\n      z-index: 1;\n      pointer-events: none;\n    }\n\n    .lh-devtools .lh-sticky-header {\n      /* The report within DevTools is placed in a container with overflow, which changes the placement of this header unless we change `position` to `sticky.` */\n      position: sticky;\n    }\n\n    .lh-sticky-header--visible {\n      display: grid;\n      grid-auto-flow: column;\n      pointer-events: auto;\n    }\n\n    /* Disable the gauge arc animation for the sticky header, so toggling display: none\n       does not play the animation. */\n    .lh-sticky-header .lh-gauge-arc {\n      animation: none;\n    }\n\n    .lh-sticky-header .lh-gauge__label {\n      display: none;\n    }\n\n    .lh-highlighter {\n      width: var(--gauge-wrapper-width);\n      height: 1px;\n      background-color: var(--highlighter-background-color);\n      /* Position at bottom of first gauge in sticky header. */\n      position: absolute;\n      grid-column: 1;\n      bottom: -1px;\n    }\n\n    .lh-gauge__wrapper:first-of-type {\n      contain: none;\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('div', 'lh-scores-wrapper');
  const el3 = dom.createElement('div', 'lh-scores-container');
  const el4 = dom.createElement('div', 'lh-pyro');
  const el5 = dom.createElement('div', 'lh-before');
  const el6 = dom.createElement('div', 'lh-after');
  el4.append(' ', el5, ' ', el6, ' ');
  el3.append(' ', el4, ' ');
  el2.append(' ', el3, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createSnippetComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-snippet');
  const el2 = dom.createElement('style');
  el2.append('\n          :root {\n            --snippet-highlight-light: #fbf1f2;\n            --snippet-highlight-dark: #ffd6d8;\n          }\n\n         .lh-snippet__header {\n          position: relative;\n          overflow: hidden;\n          padding: 10px;\n          border-bottom: none;\n          color: var(--snippet-color);\n          background-color: var(--snippet-background-color);\n          border: 1px solid var(--report-border-color-secondary);\n        }\n        .lh-snippet__title {\n          font-weight: bold;\n          float: left;\n        }\n        .lh-snippet__node {\n          float: left;\n          margin-left: 4px;\n        }\n        .lh-snippet__toggle-expand {\n          padding: 1px 7px;\n          margin-top: -1px;\n          margin-right: -7px;\n          float: right;\n          background: transparent;\n          border: none;\n          cursor: pointer;\n          font-size: 14px;\n          color: #0c50c7;\n        }\n\n        .lh-snippet__snippet {\n          overflow: auto;\n          border: 1px solid var(--report-border-color-secondary);\n        }\n        /* Container needed so that all children grow to the width of the scroll container */\n        .lh-snippet__snippet-inner {\n          display: inline-block;\n          min-width: 100%;\n        }\n\n        .lh-snippet:not(.lh-snippet--expanded) .lh-snippet__show-if-expanded {\n          display: none;\n        }\n        .lh-snippet.lh-snippet--expanded .lh-snippet__show-if-collapsed {\n          display: none;\n        }\n\n        .lh-snippet__line {\n          background: white;\n          white-space: pre;\n          display: flex;\n        }\n        .lh-snippet__line:not(.lh-snippet__line--message):first-child {\n          padding-top: 4px;\n        }\n        .lh-snippet__line:not(.lh-snippet__line--message):last-child {\n          padding-bottom: 4px;\n        }\n        .lh-snippet__line--content-highlighted {\n          background: var(--snippet-highlight-dark);\n        }\n        .lh-snippet__line--message {\n          background: var(--snippet-highlight-light);\n        }\n        .lh-snippet__line--message .lh-snippet__line-number {\n          padding-top: 10px;\n          padding-bottom: 10px;\n        }\n        .lh-snippet__line--message code {\n          padding: 10px;\n          padding-left: 5px;\n          color: var(--color-fail);\n          font-family: var(--report-font-family);\n        }\n        .lh-snippet__line--message code {\n          white-space: normal;\n        }\n        .lh-snippet__line-icon {\n          padding-top: 10px;\n          display: none;\n        }\n        .lh-snippet__line--message .lh-snippet__line-icon {\n          display: block;\n        }\n        .lh-snippet__line-icon:before {\n          content: "";\n          display: inline-block;\n          vertical-align: middle;\n          margin-right: 4px;\n          width: var(--score-icon-size);\n          height: var(--score-icon-size);\n          background-image: var(--fail-icon-url);\n        }\n        .lh-snippet__line-number {\n          flex-shrink: 0;\n          width: 40px;\n          text-align: right;\n          font-family: monospace;\n          padding-right: 5px;\n          margin-right: 5px;\n          color: var(--color-gray-600);\n          user-select: none;\n        }\n    ');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createSnippetContentComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-snippet__snippet');
  const el2 = dom.createElement('div', 'lh-snippet__snippet-inner');
  el1.append(' ', el2, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createSnippetHeaderComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-snippet__header');
  const el2 = dom.createElement('div', 'lh-snippet__title');
  const el3 = dom.createElement('div', 'lh-snippet__node');
  const el4 = dom.createElement('button', 'lh-snippet__toggle-expand');
  const el5 = dom.createElement('span', 'lh-snippet__btn-label-collapse lh-snippet__show-if-expanded');
  const el6 = dom.createElement('span', 'lh-snippet__btn-label-expand lh-snippet__show-if-collapsed');
  el4.append(' ', el5, ' ', el6, ' ');
  el1.append(' ', el2, ' ', el3, ' ', el4, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createSnippetLineComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-snippet__line');
  const el2 = dom.createElement('div', 'lh-snippet__line-number');
  const el3 = dom.createElement('div', 'lh-snippet__line-icon');
  const el4 = dom.createElement('code');
  el1.append(' ', el2, ' ', el3, ' ', el4, ' ');
  el0.append(el1);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createTopbarComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('style');
  el1.append('\n    .lh-topbar {\n      position: sticky;\n      top: 0;\n      left: 0;\n      right: 0;\n      z-index: 1000;\n      display: flex;\n      align-items: center;\n      height: var(--topbar-height);\n      background-color: var(--topbar-background-color);\n      padding: var(--topbar-padding);\n    }\n\n    .lh-topbar__logo {\n      width: var(--topbar-logo-size);\n      height: var(--topbar-logo-size);\n      user-select: none;\n      flex: none;\n    }\n\n    .lh-topbar__url {\n      margin: var(--topbar-padding);\n      text-decoration: none;\n      color: var(--report-text-color);\n      text-overflow: ellipsis;\n      overflow: hidden;\n      white-space: nowrap;\n    }\n\n    .lh-tools {\n      margin-left: auto;\n      will-change: transform;\n      min-width: var(--report-icon-size);\n    }\n    .lh-tools__button {\n      width: var(--report-icon-size);\n      height: var(--report-icon-size);\n      cursor: pointer;\n      margin-right: 5px;\n      /* This is actually a button element, but we want to style it like a transparent div. */\n      display: flex;\n      background: none;\n      color: inherit;\n      border: none;\n      padding: 0;\n      font: inherit;\n      outline: inherit;\n    }\n    .lh-tools__button svg {\n      fill: var(--tools-icon-color);\n    }\n    .lh-dark .lh-tools__button svg {\n      filter: invert(1);\n    }\n    .lh-tools__button.lh-active + .lh-tools__dropdown {\n      opacity: 1;\n      clip: rect(-1px, 194px, 242px, -3px);\n      visibility: visible;\n    }\n    .lh-tools__dropdown {\n      position: absolute;\n      background-color: var(--report-background-color);\n      border: 1px solid var(--report-border-color);\n      border-radius: 3px;\n      padding: calc(var(--default-padding) / 2) 0;\n      cursor: pointer;\n      top: 36px;\n      right: 0;\n      box-shadow: 1px 1px 3px #ccc;\n      min-width: 125px;\n      clip: rect(0, 164px, 0, 0);\n      visibility: hidden;\n      opacity: 0;\n      transition: all 200ms cubic-bezier(0,0,0.2,1);\n    }\n    .lh-tools__dropdown a {\n      color: currentColor;\n      text-decoration: none;\n      white-space: nowrap;\n      padding: 0 12px;\n      line-height: 2;\n    }\n    .lh-tools__dropdown a:hover,\n    .lh-tools__dropdown a:focus {\n      background-color: var(--color-gray-200);\n      outline: none;\n    }\n    /* save-gist option hidden in report. */\n    .lh-tools__dropdown a[data-action=\'save-gist\'] {\n      display: none;\n    }\n\n    @media screen and (max-width: 964px) {\n      .lh-tools__dropdown {\n        right: 0;\n        left: initial;\n      }\n    }\n    @media print {\n      .lh-topbar {\n        position: static;\n        margin-left: 0;\n      }\n\n      .lh-tools__dropdown {\n        display: none;\n      }\n    }\n  ');
  el0.append(el1);
  const el2 = dom.createElement('div', 'lh-topbar');
  const el3 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg', 'lh-topbar__logo');
  el3.setAttribute('viewBox', '0 0 24 24');
  const el4 = dom.createElementNS('http://www.w3.org/2000/svg', 'defs');
  const el5 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el5.setAttribute('x1', '57.456%');
  el5.setAttribute('y1', '13.086%');
  el5.setAttribute('x2', '18.259%');
  el5.setAttribute('y2', '72.322%');
  el5.setAttribute('id', 'lh-topbar__logo--a');
  const el6 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el6.setAttribute('stop-color', '#262626');
  el6.setAttribute('stop-opacity', '.1');
  el6.setAttribute('offset', '0%');
  const el7 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el7.setAttribute('stop-color', '#262626');
  el7.setAttribute('stop-opacity', '0');
  el7.setAttribute('offset', '100%');
  el5.append(' ', el6, ' ', el7, ' ');
  const el8 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el8.setAttribute('x1', '100%');
  el8.setAttribute('y1', '50%');
  el8.setAttribute('x2', '0%');
  el8.setAttribute('y2', '50%');
  el8.setAttribute('id', 'lh-topbar__logo--b');
  const el9 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el9.setAttribute('stop-color', '#262626');
  el9.setAttribute('stop-opacity', '.1');
  el9.setAttribute('offset', '0%');
  const el10 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el10.setAttribute('stop-color', '#262626');
  el10.setAttribute('stop-opacity', '0');
  el10.setAttribute('offset', '100%');
  el8.append(' ', el9, ' ', el10, ' ');
  const el11 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el11.setAttribute('x1', '58.764%');
  el11.setAttribute('y1', '65.756%');
  el11.setAttribute('x2', '36.939%');
  el11.setAttribute('y2', '50.14%');
  el11.setAttribute('id', 'lh-topbar__logo--c');
  const el12 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el12.setAttribute('stop-color', '#262626');
  el12.setAttribute('stop-opacity', '.1');
  el12.setAttribute('offset', '0%');
  const el13 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el13.setAttribute('stop-color', '#262626');
  el13.setAttribute('stop-opacity', '0');
  el13.setAttribute('offset', '100%');
  el11.append(' ', el12, ' ', el13, ' ');
  const el14 = dom.createElementNS('http://www.w3.org/2000/svg', 'linearGradient');
  el14.setAttribute('x1', '41.635%');
  el14.setAttribute('y1', '20.358%');
  el14.setAttribute('x2', '72.863%');
  el14.setAttribute('y2', '85.424%');
  el14.setAttribute('id', 'lh-topbar__logo--d');
  const el15 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el15.setAttribute('stop-color', '#FFF');
  el15.setAttribute('stop-opacity', '.1');
  el15.setAttribute('offset', '0%');
  const el16 = dom.createElementNS('http://www.w3.org/2000/svg', 'stop');
  el16.setAttribute('stop-color', '#FFF');
  el16.setAttribute('stop-opacity', '0');
  el16.setAttribute('offset', '100%');
  el14.append(' ', el15, ' ', el16, ' ');
  el4.append(' ', el5, ' ', el8, ' ', el11, ' ', el14, ' ');
  const el17 = dom.createElementNS('http://www.w3.org/2000/svg', 'g');
  el17.setAttribute('fill', 'none');
  el17.setAttribute('fill-rule', 'evenodd');
  const el18 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el18.setAttribute('d', 'M12 3l4.125 2.625v3.75H18v2.25h-1.688l1.5 9.375H6.188l1.5-9.375H6v-2.25h1.875V5.648L12 3zm2.201 9.938L9.54 14.633 9 18.028l5.625-2.062-.424-3.028zM12.005 5.67l-1.88 1.207v2.498h3.75V6.86l-1.87-1.19z');
  el18.setAttribute('fill', '#F44B21');
  const el19 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el19.setAttribute('fill', '#FFF');
  el19.setAttribute('d', 'M14.201 12.938L9.54 14.633 9 18.028l5.625-2.062z');
  const el20 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el20.setAttribute('d', 'M6 18c-2.042 0-3.95-.01-5.813 0l1.5-9.375h4.326L6 18z');
  el20.setAttribute('fill', 'url(#lh-topbar__logo--a)');
  el20.setAttribute('fill-rule', 'nonzero');
  el20.setAttribute('transform', 'translate(6 3)');
  const el21 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el21.setAttribute('fill', '#FFF176');
  el21.setAttribute('fill-rule', 'nonzero');
  el21.setAttribute('d', 'M13.875 9.375v-2.56l-1.87-1.19-1.88 1.207v2.543z');
  const el22 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el22.setAttribute('fill', 'url(#lh-topbar__logo--b)');
  el22.setAttribute('fill-rule', 'nonzero');
  el22.setAttribute('d', 'M0 6.375h6v2.25H0z');
  el22.setAttribute('transform', 'translate(6 3)');
  const el23 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el23.setAttribute('fill', 'url(#lh-topbar__logo--c)');
  el23.setAttribute('fill-rule', 'nonzero');
  el23.setAttribute('d', 'M6 6.375H1.875v-3.75L6 0z');
  el23.setAttribute('transform', 'translate(6 3)');
  const el24 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el24.setAttribute('fill', 'url(#lh-topbar__logo--d)');
  el24.setAttribute('fill-rule', 'nonzero');
  el24.setAttribute('d', 'M6 0l4.125 2.625v3.75H12v2.25h-1.688l1.5 9.375H.188l1.5-9.375H0v-2.25h1.875V2.648z');
  el24.setAttribute('transform', 'translate(6 3)');
  el17.append(' ', el18, ' ', el19, ' ', el20, ' ', el21, ' ', el22, ' ', el23, ' ', el24, ' ');
  el3.append(' ', el4, ' ', el17, ' ');
  const el25 = dom.createElement('a', 'lh-topbar__url');
  el25.setAttribute('href', '');
  el25.setAttribute('target', '_blank');
  el25.setAttribute('rel', 'noopener');
  const el26 = dom.createElement('div', 'lh-tools');
  const el27 = dom.createElement('button', 'lh-tools__button');
  el27.setAttribute('id', 'lh-tools-button');
  el27.setAttribute('title', 'Tools menu');
  el27.setAttribute('aria-label', 'Toggle report tools menu');
  el27.setAttribute('aria-haspopup', 'menu');
  el27.setAttribute('aria-expanded', 'false');
  el27.setAttribute('aria-controls', 'lh-tools-dropdown');
  const el28 = dom.createElementNS('http://www.w3.org/2000/svg', 'svg');
  el28.setAttribute('width', '100%');
  el28.setAttribute('height', '100%');
  el28.setAttribute('viewBox', '0 0 24 24');
  const el29 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el29.setAttribute('d', 'M0 0h24v24H0z');
  el29.setAttribute('fill', 'none');
  const el30 = dom.createElementNS('http://www.w3.org/2000/svg', 'path');
  el30.setAttribute('d', 'M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z');
  el28.append(' ', el29, ' ', el30, ' ');
  el27.append(' ', el28, ' ');
  const el31 = dom.createElement('div', 'lh-tools__dropdown');
  el31.setAttribute('id', 'lh-tools-dropdown');
  el31.setAttribute('role', 'menu');
  el31.setAttribute('aria-labelledby', 'lh-tools-button');
  const el32 = dom.createElement('a', 'lh-report-icon lh-report-icon--print');
  el32.setAttribute('role', 'menuitem');
  el32.setAttribute('tabindex', '-1');
  el32.setAttribute('href', '#');
  el32.setAttribute('data-i18n', 'dropdownPrintSummary');
  el32.setAttribute('data-action', 'print-summary');
  const el33 = dom.createElement('a', 'lh-report-icon lh-report-icon--print');
  el33.setAttribute('role', 'menuitem');
  el33.setAttribute('tabindex', '-1');
  el33.setAttribute('href', '#');
  el33.setAttribute('data-i18n', 'dropdownPrintExpanded');
  el33.setAttribute('data-action', 'print-expanded');
  const el34 = dom.createElement('a', 'lh-report-icon lh-report-icon--copy');
  el34.setAttribute('role', 'menuitem');
  el34.setAttribute('tabindex', '-1');
  el34.setAttribute('href', '#');
  el34.setAttribute('data-i18n', 'dropdownCopyJSON');
  el34.setAttribute('data-action', 'copy');
  const el35 = dom.createElement('a', 'lh-report-icon lh-report-icon--download');
  el35.setAttribute('role', 'menuitem');
  el35.setAttribute('tabindex', '-1');
  el35.setAttribute('href', '#');
  el35.setAttribute('data-i18n', 'dropdownSaveHTML');
  el35.setAttribute('data-action', 'save-html');
  const el36 = dom.createElement('a', 'lh-report-icon lh-report-icon--download');
  el36.setAttribute('role', 'menuitem');
  el36.setAttribute('tabindex', '-1');
  el36.setAttribute('href', '#');
  el36.setAttribute('data-i18n', 'dropdownSaveJSON');
  el36.setAttribute('data-action', 'save-json');
  const el37 = dom.createElement('a', 'lh-report-icon lh-report-icon--open');
  el37.setAttribute('role', 'menuitem');
  el37.setAttribute('tabindex', '-1');
  el37.setAttribute('href', '#');
  el37.setAttribute('data-i18n', 'dropdownViewer');
  el37.setAttribute('data-action', 'open-viewer');
  const el38 = dom.createElement('a', 'lh-report-icon lh-report-icon--open');
  el38.setAttribute('role', 'menuitem');
  el38.setAttribute('tabindex', '-1');
  el38.setAttribute('href', '#');
  el38.setAttribute('data-i18n', 'dropdownSaveGist');
  el38.setAttribute('data-action', 'save-gist');
  const el39 = dom.createElement('a', 'lh-report-icon lh-report-icon--dark');
  el39.setAttribute('role', 'menuitem');
  el39.setAttribute('tabindex', '-1');
  el39.setAttribute('href', '#');
  el39.setAttribute('data-i18n', 'dropdownDarkTheme');
  el39.setAttribute('data-action', 'toggle-dark');
  el31.append(' ', el32, ' ', el33, ' ', el34, ' ', el35, ' ', el36, ' ', el37, ' ', el38, ' ', el39, ' ');
  el26.append(' ', el27, ' ', el31, ' ');
  el2.append(' ', ' ', el3, ' ', el25, ' ', el26, ' ');
  el0.append(el2);
  return el0;
}

/**
 * @param {DOM} dom
 */
function createWarningsToplevelComponent(dom) {
  const el0 = dom.document().createDocumentFragment();
  const el1 = dom.createElement('div', 'lh-warnings lh-warnings--toplevel');
  const el2 = dom.createElement('p', 'lh-warnings__msg');
  const el3 = dom.createElement('ul');
  el1.append(' ', el2, ' ', el3, ' ');
  el0.append(el1);
  return el0;
}


/** @typedef {'3pFilter'|'audit'|'categoryHeader'|'chevron'|'clump'|'crc'|'crcChain'|'elementScreenshot'|'envItem'|'footer'|'fraction'|'gauge'|'gaugePwa'|'heading'|'metric'|'metricsToggle'|'opportunity'|'opportunityHeader'|'scorescale'|'scoresWrapper'|'snippet'|'snippetContent'|'snippetHeader'|'snippetLine'|'topbar'|'warningsToplevel'} ComponentName */
/**
 * @param {DOM} dom
 * @param {ComponentName} componentName
 * @return {DocumentFragment}
 */
export function createComponent(dom, componentName) {
  switch (componentName) {
    case '3pFilter': return create3pFilterComponent(dom);
    case 'audit': return createAuditComponent(dom);
    case 'categoryHeader': return createCategoryHeaderComponent(dom);
    case 'chevron': return createChevronComponent(dom);
    case 'clump': return createClumpComponent(dom);
    case 'crc': return createCrcComponent(dom);
    case 'crcChain': return createCrcChainComponent(dom);
    case 'elementScreenshot': return createElementScreenshotComponent(dom);
    case 'envItem': return createEnvItemComponent(dom);
    case 'footer': return createFooterComponent(dom);
    case 'fraction': return createFractionComponent(dom);
    case 'gauge': return createGaugeComponent(dom);
    case 'gaugePwa': return createGaugePwaComponent(dom);
    case 'heading': return createHeadingComponent(dom);
    case 'metric': return createMetricComponent(dom);
    case 'metricsToggle': return createMetricsToggleComponent(dom);
    case 'opportunity': return createOpportunityComponent(dom);
    case 'opportunityHeader': return createOpportunityHeaderComponent(dom);
    case 'scorescale': return createScorescaleComponent(dom);
    case 'scoresWrapper': return createScoresWrapperComponent(dom);
    case 'snippet': return createSnippetComponent(dom);
    case 'snippetContent': return createSnippetContentComponent(dom);
    case 'snippetHeader': return createSnippetHeaderComponent(dom);
    case 'snippetLine': return createSnippetLineComponent(dom);
    case 'topbar': return createTopbarComponent(dom);
    case 'warningsToplevel': return createWarningsToplevelComponent(dom);
  }
  throw new Error('unexpected component: ' + componentName);
}
